Copilot Coding Agent documentation (#8653)
microsoft/vscode-docs: docs/copilot/copilot-coding-agent.md

* **VS Code setup**: Install the [GitHub Pull Requests extension](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github)

**Optional**: Enable the experimental setting `setting(githubPullRequests.codingAgent.uiIntegration)` to show a **Delegate to coding agent** button in Copilot Chat for easier task delegation.

> [!TIP]
> If you don't have Copilot access yet, you can sign up for the [Copilot Free plan](https://github.com/features/copilot/plans) to get a monthly limit of interactions.

## Assign work to Copilot coding agent in VS Code

### Method 1: Assign issues to Copilot

You can trigger Copilot coding agent by assigning a GitHub issue to Copilot, similar to how you assign an issue to a team member. Copilot coding agent automatically analyzes the issue and starts working on it.

1. In the **GitHub Pull Requests** view, navigate to the **Issues** section

1. Find the issue you want to assign to Copilot

1. Right-click the issue and select **Assign to Copilot** or select **Assign** and then select `@copilot`

1. The agent will begin working on the issue in the background

> > [!TIP]

Author: Great feedback! Moved to previous step and indented.

---
ContentId: f8b9e2a4-7c1d-4f5e-9a8b-3d2e1f0c6789
DateApproved: 07/09/2025
MetaDescription: Learn how to interact with the GitHub Copilot coding agent in VS Code to autonomously implement features and fix bugs in the background.
MetaSocialImage: images/shared/github-copilot-social.png
---
# GitHub Copilot coding agent

[GitHub Copilot Coding Agent](https://docs.github.com/en/copilot/concepts/about-copilot-coding-agent) is an autonomous AI developer that works independently in the background to complete development tasks. You can assign GitHub issues to Copilot or delegate tasks in chat, and the agent will work autonomously to implement features, fix bugs, and make changes across your repository using its own isolated development environment.

![GIF showing how to assign an issue to Copilot coding agent from within VS Code.](images/copilot-coding-agent/assign-to-copilot-gif.gif)

> [!NOTE]
> Copilot coding agent is in public preview and subject to change. During the preview, use of the feature is subject to [GitHub Pre-release License Terms](https://docs.github.com/en/site-policy/github-terms/github-pre-release-license-terms).

## What Copilot coding agent can do

> Copilot coding agent is a GitHub-hosted autonomous agent that works independently in an isolated environment to complete development tasks, such as:

Reviewer: The intro looks better - thanks! Wonder if we should be explicit that those isolated development environments run on GitHub? One might still assume that coding agent spins up a local dev env/container.

* **Fix bugs and implement features** across multiple files
* **Improve test coverage** and run automated checks
* **Refactor code and address technical debt**
* **Create pull requests** and work independently of your local environment

This is different from the local [agent mode](/docs/copilot/chat/chat-agent-mode.md) in VS Code, which runs locally in your editor and requires your active participation during the coding session.

## How it works

The Copilot coding agent workflow:

> 

Reviewer: The information is there for now, so no rush to add a diagram.

1. **Pull request**: The agent creates a pull request with the implementation
1. **Review**: You review the changes and can request modifications through PR comments
1. **Iteration**: The agent responds to feedback and updates the implementation

## Prerequisites

Before you can use Copilot coding agent, you need:

* **GitHub Copilot subscription**: Available with Copilot Pro, Pro+, Business, or Enterprise plans
* **Write access**: You must have write permissions to the repository
* **Enable the agent**: Copilot coding agent must be enabled for your account or organization
* **VS Code setup**: Install the [GitHub Pull Requests extension](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github)

**Optional**: Enable the experimental setting `setting(githubPullRequests.codingAgent.uiIntegration)` to show a **Delegate to coding agent** button in Copilot Chat for easier task delegation.

> [!TIP]
> If you don't have Copilot access yet, you can sign up for the [Copilot Free plan](https://github.com/features/copilot/plans) to get a monthly limit of interactions.

## Getting started in VS Code

### Method 1: Assign issues to Copilot

> 

Author: Love this idea! Added.

1. In the **GitHub Pull Requests** view, navigate to the **Issues** section

1. Find the issue you want to assign to Copilot

1. Right-click the issue and select **Assign to Copilot** or click the assign button and select `@copilot`

1. The agent will begin working on the issue in the background

   ![Screenshot showing the GitHub Pull Requests view, highlighting the assign to Copilot action, and the PR query for work assigned to Copilot.](images/copilot-coding-agent/github-pull-request-coding-agent.png)

### Method 2: Delegate from Copilot Chat

> 

Author: Great feedback, added!

> 1. Open Copilot Chat (`kb(workbench.action.chat.open)`)

Reviewer: ```suggestion
1. Open the Chat view in VS Code (`kb(workbench.action.chat.open)`)
```

1. Have a conversation about the feature or change you want to implement

> 1. When ready, delegate to the agent using one of these methods:

Reviewer: ```suggestion
1. When ready, delegate to the agent by using one of these methods:
```

Author: Thanks!

   **Use the delegate button (Experimental)**

   Enable the experimental setting `setting(githubPullRequests.codingAgent.uiIntegration)` to show a **Delegate to coding agent** button in the Chat view for repositories that have the agent enabled. Select this button to hand off your current chat context to the coding agent.

   <video src="images/copilot-coding-agent/delegate-to-coding-agent.mp4" title="Video showing how to delegate to coding agent from VS Code chat." controls poster="images/copilot-coding-agent/delegate-to-coding-agent-poster.png"></video>

   **Use the #copilotCodingAgent tool**

   You can also reference the tool directly in your prompt to ask Copilot to continue a local change in the background. This tool automatically pushes pending changes to a remote branch and initiates a coding agent session:

   ![Screenshot showing handing off a session to Copilot coding agent](images/copilot-coding-agent/coding-agent-start.png)

1. The agent will create a pull request and begin implementing the discussed changes

> ## Tracking agent progress

Reviewer: ```suggestion
## Track agent progress
```

### Monitor work in VS Code

The GitHub Pull Requests extension provides a dedicated **Copilot on My Behalf** section that shows:

* All active Copilot coding agent sessions
* Pull requests created by the agent
* Progress status for each task
* Numeric badges indicating new changes or updates

![Screenshot showing status of multiple coding agent pull requests](images/copilot-coding-agent/coding-agent-status.png)

> 

Reviewer: Include a sentence that you can also use this for work that you assigned through GH.com

Author: Thanks, added a tip!

### View detailed session logs

1. In the Pull Requests view, find your agent's work under **Copilot on My Behalf**

1. Select **View Session** to see a detailed log of everything the agent did:
   * Commands executed
   * Files modified
   * Tests run
   * Decision-making process

   ![Screenshot showing the session log of a coding agent session.](images/copilot-coding-agent/coding-agent-session-log.png)

> [!TIP]

> > When working with pull requests created by the coding agent, the `#activePullRequest` tool is automatically attached to chat, allowing you to maintain context and continue working on the pull request with access to the coding agent session information.

Reviewer: This is not entirely clear to me. Can you elaborate which scenarios this enables? Does this mean I can intervene on the PR, or do work on the side and the coding agent picks up my changes?

### Cancel a running session

If you need to stop the agent, you can stay in VS Code and use the **Cancel coding agent** button on the PR overview page.

You can also cancel a session from GitHub.com:
1. Go to your GitHub repository on GitHub.com
1. Navigate to the **Actions** tab
1. Find the running Copilot Coding Agent workflow
1. Select **Cancel workflow**

> ## Reviewing and iterating

Reviewer: ```suggestion
## Review and iterate
```

> ### When the agent completes work

Reviewer: ```suggestion
### Work completion
```

> The Copilot coding agent will:

Reviewer: The intro sentence could be expanded a bit more to stand by itself. Right now, it depends on the header to provide sufficient context.

* Create a pull request with all changes
* Assign the PR to you for review
* Request you as a reviewer
* Include a detailed description explaining the implementation
* Add screenshots when applicable (for UI changes)

![Screenshot showing a pull request from Copilot coding agent displayed in VS Code with an included screenshot of the implemented feature.](images/copilot-coding-agent/draft-with-screenshot.png)

> ### Providing feedback

Reviewer: ```suggestion
### Provide feedback
```

### Can I extend Copilot coding agent with external tools?

For advanced scenarios, you can extend Copilot coding agent with Model Context Protocol (MCP) servers to give it access to:

* External databases
* Cloud services
* APIs and third-party integrations
* Custom development tools

Learn more about [extending Copilot coding agent with MCP](https://docs.github.com/en/copilot/using-github-copilot/coding-agent/extending-copilot-coding-agent-with-mcp).

### What are the current limitations?

* **Cross-repository changes**: Can only work within the repository where the issue is assigned
* **Multiple PRs per task**: Opens exactly one pull request per assigned task
* **Existing PR modifications**: Cannot work on pull requests it didn't create

For detailed information about limitations, compatibility, and usage costs, see the [GitHub Copilot Coding Agent documentation](https://docs.github.com/en/copilot/using-github-copilot/coding-agent).

## Getting help

### Hands-on practice

> Try the [Expand your team with Copilot coding agent](https://github.com/skills/expand-your-team-with-copilot/) Skills exercise for practical experience.

Reviewer: Shouldn't this be a next step item instead of a standalone section?

## Next steps

* Enable Copilot coding agent by following the [GitHub setup guide](https://docs.github.com/en/copilot/using-github-copilot/coding-agent/enabling-copilot-coding-agent)

> * Try [local agent mode](/docs/copilot/chat/chat-agent-mode.md) in VS Code for immediate, interactive coding assistance

Reviewer: ```suggestion
* Try [agent mode in VS Code](/docs/copilot/chat/chat-agent-mode.md) for immediate, interactive coding assistance
```

> * Learn about [Copilot Chat](/docs/copilot/chat/copilot-chat.md) for conversational AI help

Reviewer: ```suggestion
* Learn about [using chat in VS Code](/docs/copilot/chat/copilot-chat.md) for conversational AI help
```
